ls1intum/Artemis
# [Chore] Refactor Test Runs

### Checklist

### Motivation and Context
In #2738 we introduced a new boolean flag for the participation: `test_run`. This PR is a follow-up to refactor the test run feature, so that we only use this flag (together with `studentExam.test_run`) to distinguish test run student exams, participations, submissions, and results. Additionally, through the new changes, we can more closely align the test runs to real exams, making them more robust to changes and providing the instructors confidence that everything will work out during the exam. 

**The following issues are addressed in this PR:** 

**Server:** 

1. The repositories still use the old testRun identification condition: `result.assessor = participation.user` in most queries.
2. We do not need to mark the submissions of test runs anymore by adding a dummy result with the condition: `result.assessor = participation.user`
3. Test run quizzes are never evaluated and never display the users result.

**Client:** 
1. After the test run conduction, the results are immediately displayed leading to confusing notices, because there are no assessments yet, and quiz exercise information is hidden during the conduction. 
2. The second correction currently breaks the complaint functionality of test runs. (This happens because there is no second correction for test runs, and it tries to access a result which is not present)
3. In #2738 we delete all test runs, if the user clicks on "Assess all unsubmitted exams and empty submissions" after the exam conduction. This however is not mentioned to the user. 
4. The status indicator of the assessments per exercise is disabled for test runs, however the client still calls  `getStatsForTutors` unnecessarily.
5. When assessing the different test run participation for exercises, in the exercise assessment view the user cannot distinguish between real assessments and test run assessments. 
6. In the "create test run" modal, the user has to manually select every exercise for each exercise group, even if there is only one option.

### Description

- [x] Remove the previous testRun identification `result.assessor != participation.user` from all queries, and replace with `p.testRun = FALSE OR p.testRun IS NULL`
- [x] Remove the old `markSubmissionsOfTestRuns` method.
- [x] After submitting a test run, immediately evaluate the quiz participation (but ignore it from the quiz statistics).
- [x] Rework the test run submission assessment call to use the `lockSubmission` method of `SubmissionService` through a wrapper in order to not alter the code.
- [x] Only display the results of the test run when the user explicitly accesses the summary. _(similar to the real exam participation where the results are published at a later date, i.e, not immediately after the conduction.)_
- [x] Because there is no second correction for test runs, it does not make sense to display the "enable second correction button". Disable the button for test runs and display a notice to the user: "Second correction does not apply to test runs", if the exam has more than one correction round.
- [x] Adapt the notice to delete the test run in the "Test Run Management" view to let the instructor know that the test runs will be automatically deleted once he assesses all unsubmitted exams and empty submissions.
- [x] Do not call  `getStatsForTutors` for Test runs, as the status indicator is disabled.
- [x] Add "Test Run" to the header of the exercise assessment in order to help the user distinguish between assessing real submissions and test run submissions.
- [x] Automatically select the exercise for every exercise group, if there is only one in the "create test run" modal.  

**Unrelated changes:** 

- [x] Rename left-over "tutor-dashboard" calls to assessment-dashboard
- [x] Tutor leaderboards and side panels are disabled for exams. Therefore, do not load the `TutorLeaderboardEntries` (unneccesary database calls). 

**WARNING:** Deleting old test runs will fail because the `AssessmentType.TEST_RUN` was deleted from the server. In order for this to work, you must run the following query on your database:

> Update result set assessment_type = 'MANUAL' where assessment_type = 'TEST_RUN'

### Steps for Testing

**Test Runs**
1. Create an exam with all exercise types
2. Create a test run. If you only have one exercise per exercise group, it should be automatically selected. 
3. Change the working time vie Manage Working Time from the Test Run Management view.
4. Start the test run and submit. You should not see "No graded results", on top of exercises.
5. Assess the test run submissions. All assessment pages should include "Test Run" in the title.
6. Go to the summary page, you should see your results (including the quiz result)
7. Complain about the assessment.
8. Respond to the complaints.
9. View the complaints in the summary.
10. Change the exam to have 2 correction rounds.
11. Click on assess your test runs. It should now display: `Hint: The second correction does not apply to test runs.`
12. Go to the scores page, no test run should be included here.

**Normal exam**
_Verify that everything works as expected, there should be no changes in behaviour_
1. Register users for the exam and create the student exams & prepare exercises.
2. Change the working time of the student (make sure no console errors are thrown!).
3. Participate in the exam.
4. Assess all unsubmitted exams and empty submissions from the "student exams" view.
5. Go to Test Runs, and verify that all test runs were deleted. 
6. Assess the exam with 2 correction rounds enabled
7. Publish the results and view them 
8. Complain about the results
9. Respond to complaints
10. View the scores page

 

### Test Coverage
SubmissionResource: 91%
StudentExamResource: 92%

ExamService: 91%
StudentExamService: 90%
SubmissionService: 92%
ParticipationService: 79%
ExamQuizSubmissionService: 68%

### Screenshots
![image](https://user-images.githubusercontent.com/15110960/106629207-e4902700-657a-11eb-96fd-3f849f454421.png)
![image](https://user-images.githubusercontent.com/15110960/106629273-f376d980-657a-11eb-8817-d86ba2d9f529.png)
![image](https://user-images.githubusercontent.com/15110960/106629376-0f7a7b00-657b-11eb-8fbe-6ea91df724b0.png)
![image](https://user-images.githubusercontent.com/15110960/106735877-0e962780-6615-11eb-92fe-81bc94a23546.png)
